### PR TITLE
[Merged by Bors] - feat: define restrictions of perfect pairings

### PR DESCRIPTION
**Pull request by ocfnash**
The main results are:
 * `PerfectPairing.restrict`
 * `PerfectPairing.restrictScalars`
 * `PerfectPairing.restrictScalarsField`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/PerfectPairing.lean [hunk 1]
```diff
@@ -45,6 +49,34 @@ variable {R M N}
 
 namespace PerfectPairing
 
+/-- If the coefficients are a field, and one of the spaces is finite-dimensional, it is sufficient
+to check only injectivity instead of bijectivity of the bilinear form. -/
+def mkOfFiniteDimensional {K V W : Type*}
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/PerfectPairing.lean:54]: Hmm, I'm wondering whether
```suggestion
def mkOfInjective {K V W : Type*}
```
might fit a tiny bit better. Do you have a preference?

File: Mathlib/LinearAlgebra/PerfectPairing.lean [hunk 2]
```diff
@@ -129,6 +161,135 @@ include p in
 theorem reflexive_right : IsReflexive R N :=
   p.flip.reflexive_left
 
+private lemma restrict_aux
+    {M' N' : Type*} [AddCommGroup M'] [Module R M'] [AddCommGroup N'] [Module R N']
+    (i : M' →ₗ[R] M) (hi : Injective i)
+    (j : N' →ₗ[R] N) (hj : Injective j)
+    (hM : IsCompl (LinearMap.range i) ((LinearMap.range j).dualAnnihilator.map p.toDualLeft.symm))
+    (hN : IsCompl (LinearMap.range j) ((LinearMap.range i).dualAnnihilator.map p.toDualRight.symm)):
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/PerfectPairing.lean:169]:
```suggestion
    (hN : IsCompl (LinearMap.range j) ((LinearMap.range i).dualAnnihilator.map p.toDualRight.symm)) :
```

File: Mathlib/LinearAlgebra/PerfectPairing.lean [hunk 3]
```diff
@@ -129,6 +161,135 @@ include p in
...
+    simpa [← hF, ← LinearMap.congr_fun hm' (j n)]
+
+/-- The restriction of a perfect pairing to submodules (expressed as injections to provide
+definitional control). -/
+@[simps]
+def restrict {M' N' : Type*} [AddCommGroup M'] [Module R M'] [AddCommGroup N'] [Module R N']
+    (i : M' →ₗ[R] M) (hi : Injective i)
+    (j : N' →ₗ[R] N) (hj : Injective j)
+    (hM : IsCompl (LinearMap.range i) ((LinearMap.range j).dualAnnihilator.map p.toDualLeft.symm))
+    (hN : IsCompl (LinearMap.range j) ((LinearMap.range i).dualAnnihilator.map p.toDualRight.symm)):
```
Comment (jcommelin) [inline Mathlib/LinearAlgebra/PerfectPairing.lean:199]:
```suggestion
    (hN : IsCompl (LinearMap.range j) ((LinearMap.range i).dualAnnihilator.map p.toDualRight.symm)) :
```

File: Mathlib/LinearAlgebra/PerfectPairing.lean (hunk 1 above)
Comment (ocfnash) [reply Mathlib/LinearAlgebra/PerfectPairing.lean:54]: Oh yeah, this is a much better name; thanks.

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**